src/AbstractNormalizedEventManager.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Dhii\WpEvents;
+
+use Psr\EventManager\EventInterface;
+
+/**
+ * Common functionality for WordPress event managers that use callback wrappers
+ * to normalize event handler parameters.
+ *
+ * @since [*next-version*]
+ */
+abstract class AbstractNormalizedEventManager extends AbstractWpEventManager
+{
+    /**
+     * Gets an event callback wrapper.
+     *
+     * @since [*next-version*]
+     * @see _createCallbackWrapper()
+     *
+     * @param string   $name     Name of the event.
+     * @param callable $callback Handler of the event.
+     *
+     * @return \Closure The wrapper.
+     */
+    protected function _getHandlerWrapper($name, $callback)
+    {
+        return $this->_createHandlerWrapper($name, $callback);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _attach($name, $callback, $priority = self::DEFAULT_PRIORITY, $numArgs = 1)
+    {
+        $handler = $this->_getHandlerWrapper($name, $callback);
+        $this->_addHook($name, $handler, $priority, $numArgs);
+
+        return $this;
+    }
+
+    /**
+     * Creates a wrapper for an event handler, containing and replacing it.
+     *
+     * If the wrapper receives an {@see EventInterface}, it will use that and assume it was
+     * triggered or normalized earlier by the event manager.
+     * Otherwise, it will assume that regular parameters are passed, such as
+     * what WordPress does, and will create an event object from that.
+     * Also allows events to stop propagation.
+     *
+     * @since [*next-version*]
+     *
+     * @param string   $name     The name of the event, for which a wrapper is created.
+     * @param callable $callback The callback to wrap.
+     *
+     * @return \Closure The wrapper.
+     */
+    protected function _createHandlerWrapper($name, $callback)
+    {
+        $me = $this;
+
+        return function () use ($name, &$callback, &$me) {
+            $args  = \func_get_args();
+            $event = \count($args) && $args[0] instanceof EventInterface
+                    ? $args[0]
+                    : $me->createEvent($name, $args);
+            /* @var $event \Dhii\WpEvents\Event */
+            if (!$event->isPropagationStopped()) {
+                \call_user_func_array($callback, array($event));
+            }
+
+            return $event;
```

Comment:
I love how you handled this using closure returns. And support for the propagation flag is nice bonus. Truly amazing! :)

If I may add however, this won't allow the manager to be used for WordPress filters.

As an example, let's say I want to remove the _Quick Edit_ row action from all posts. The `post_row_actions` WordPress filter can be used to return an array of row actions. If I use an event manager that uses this closure as a wrapper for my handler, the `Event` instance will be returned, even if my handler complies to standard and depends on a `$event` argument.

Consider:

```php
$manager->attach('post_row_actions', function($event) {
    // Get params array and then the row actions by reference
    $params     = $event->getParams();
    $rowActions = &$params[0];
    // Remove "Quick Edit"
    unset($rowActions['inline hide-if-no-js']);
    // Set new params
    $event->setParams($params);
}
```

The above adheres to the standard. Action has been taken in the handler and the `Event`'s data has changed. This is all allowed. However, WordPress requires a _result_, and currently the wrapper only returns the _Event_ instance.

Since we cannot count on WordPress adhering to the PSR-14 event manager standard, if we wish to make the event manager compatible with attaching handlers to WordPress hooks, I recommend the following changes inside the wrapper:

Detect if the event is a WordPress event. This is easy - WordPress-triggered hooks do not pass `Event` instances.

```php
$isWpEvent = !\count($args) || !$args[0] instanceof EventInterface;
$event     = $isWpEvent
    ? $me->createEvent($name, $args)
    : $args[0];
```

So far, nothing has changed except for the addition of the flag `$isWpEvent`. This can now be used at `return`-time:

```
return $isWpEvent
    ? $event->getParam(0)
    : $event;
```

Since WordPress treats the first-argument as the "value to be modified by filters", if a WordPress event is detected, the first event parameter is returned instead of the `Event` instance. A more standardized way would be to settle on a param "key", such as `"result"` or perhaps something less intrusive like `"_filter"`.

This is of course like I said, if we wish to make the event manager capable of attaching handlers to WordPress-triggered hooks.